File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/vehicle/GravityForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.vehicle;

import com.github.aadvorak.artilleryonline.battle.calculations.BodyForce;
import com.github.aadvorak.artilleryonline.battle.calculations.ForceCalculator;
import com.github.aadvorak.artilleryonline.battle.calculations.VehicleCalculations;
import com.github.aadvorak.artilleryonline.battle.common.Constants;
import com.github.aadvorak.artilleryonline.battle.common.Contact;
import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class GravityForceCalculator implements ForceCalculator<
        VehicleSpecs,
        VehiclePreCalc,
        VehicleConfig,
        VehicleState,
        VehicleModel,
        VehicleCalculations> {

    private static final String FORCE_DESCRIPTION = "Gravity";

    @Override
    public List<BodyForce> calculate(VehicleCalculations calculations, BattleModel battleModel) {
        var roomGravityAcceleration = battleModel.getRoom().getSpecs().getGravityAcceleration();
        var mass = calculations.getMass();
        var forces = new ArrayList<BodyForce>();
        var groundContacts = calculations.getAllGroundContacts();
        var comX = calculations.getPosition().getX();
        var leftContacts = groundContacts.stream()
                .filter(contact -> contact.position().getX() < comX)
                .collect(Collectors.toSet());
        var rightContacts = groundContacts.stream()
                .filter(contact -> contact.position().getX() > comX)
                .collect(Collectors.toSet());
        if (leftContacts.isEmpty() || rightContacts.isEmpty()) {
            forces.add(BodyForce.atCOM(
                    new Force().setX(0.0).setY(-roomGravityAcceleration * mass),
                    FORCE_DESCRIPTION
            ));
        } else {
            var leftContact = getFarthest(leftContacts, comX);
            var rightContact = getFarthest(rightContacts, comX);
            addGroundForce(forces, leftContact, mass, roomGravityAcceleration);
            addGroundForce(forces, rightContact, mass, roomGravityAcceleration);
        }
        return forces;
    }

    private void addGroundForce(
            List<BodyForce> forces,
            Contact contact,
            double mass,
            double roomGravityAcceleration
    ) {
        final var maxFriction = 1.0;
        var forceModule = Math.max(
                Math.abs(roomGravityAcceleration * Math.sin(contact.angle())) * mass / 2 - maxFriction,
                0.0
        );
        if  (forceModule > 0.0) {
            forces.add(BodyForce.atCOM(
                    new Force()
                            .setX(-forceModule * Math.sin(contact.angle()))
                            .setY(-forceModule * Math.cos(contact.angle())),
                    FORCE_DESCRIPTION
            ));
        }
    }

    private Contact getFarthest(Set<Contact> contacts, double comX) {
        var iterator = contacts.iterator();
        var farthest = iterator.next();
        var farthestDistance = xDistance(farthest, comX);
        while (iterator.hasNext()) {
            var contact = iterator.next();
            var xDistance = xDistance(contact, comX);
            if (xDistance > farthestDistance) {
                farthest = contact;
                farthestDistance = xDistance;
            }
        }
        return farthest;
    }

    private double xDistance(Contact contact, double comX) {
        return Math.abs(contact.position().getX() - comX);
    }
}

File: frontend/playground/battle/calculator/common/gravity-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import type {BodyCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import type {Contact} from "~/playground/data/common";
import {BodyUtils} from "~/playground/utils/body-utils";

export class GravityForceCalculator implements ForceCalculator<BodyCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Gravity'

  calculate(calculations: BodyCalculations, battleModel: BattleModel): BodyForce[] {
    const roomGravityAcceleration = battleModel.room.specs.gravityAcceleration
    const mass = calculations.model.preCalc.mass
    const forces: BodyForce[] = []

    const comX = calculations.model.state.position.x
    const contactsArray = BodyUtils.getAllGroundContacts(calculations)

    const leftContacts = contactsArray.filter(c => c.position.x < comX)
    const rightContacts = contactsArray.filter(c => c.position.x > comX)

    if (leftContacts.length === 0 || rightContacts.length === 0) {
      forces.push(BodyForce.atCOM(
          {x: 0, y: -roomGravityAcceleration * mass},
          GravityForceCalculator.FORCE_DESCRIPTION
      ))
    } else {
      const leftContact = this.getFarthest(leftContacts, comX)
      const rightContact = this.getFarthest(rightContacts, comX)
      this.addGroundForce(forces, leftContact, mass, roomGravityAcceleration)
      this.addGroundForce(forces, rightContact, mass, roomGravityAcceleration)
    }

    return forces
  }

  private addGroundForce(
      forces: BodyForce[],
      contact: Contact,
      mass: number,
      roomGravityAcceleration: number
  ): void {
    const maxFriction = 1.0
    const forceModule = Math.max(
        Math.abs(roomGravityAcceleration * Math.sin(contact.angle)) * mass / 2 - maxFriction,
        0.0
    )

    if (forceModule > 0) {
      forces.push(BodyForce.atCOM(
          {
            x: -forceModule * Math.sin(contact.angle),
            y: -forceModule * Math.cos(contact.angle)
          },
          GravityForceCalculator.FORCE_DESCRIPTION
      ))
    }
  }

  private getFarthest(contacts: Contact[], comX: number): Contact {
    let farthest = contacts[0]
    let farthestDistance = this.xDistance(farthest, comX)

    for (const contact of contacts) {
      const currentDistance = this.xDistance(contact, comX)
      if (currentDistance > farthestDistance) {
        farthest = contact
        farthestDistance = currentDistance
      }
    }

    return farthest
  }

  private xDistance(contact: Contact, comX: number): number {
    return Math.abs(contact.position.x - comX)
  }
}
